Misc changes to StableMIR required to Kani use case.

celinval opened the pull request:
First, I wanted to say that I can split this review into multiple if it makes reviewing easier. I bundled them up, since I've been testing them together (See https://github.com/rust-lang/project-stable-mir/pull/51 for the set of more thorough checks).

So far, this review includes 3 commits:

1. Add more APIs and fix `Instance::body`
    - Add more APIs to retrieve information about types.
    - Add a few more instance resolution options. For the drop shim, we return None if the drop body is empty. Not sure it will be enough.
    - Make `Instance::body()` return an Option<Body>, since not every instance might have an available body. For example, foreign instances, virtual instances, dependencies.
2. Fix a bug on MIRVisitor
    - We were not iterating over all local variables due to a typo.
3. Add more SMIR internal impl and callback return value
    - In cases like Kani, we will invoke the rustc_internal run command directly for now. It would be handly to be able to have a callback that can return a value.
    - We also need extra methods to convert stable constructs into internal ones, so we can break down the transition into finer grain commits.
    - For the internal implementation of Region, we're always returning `ReErased` for now.

celinval committed:
Add more SMIR internal impl and callback return value

In cases like Kani, we will invoke the rustc_internal run command
directly for now. It would be handly to be able to have a callback
that can return a value.

We also need extra methods to convert stable constructs into internal
ones, so we can break down the transition into finer grain commits.

celinval committed:
Fix bug on MIRVisitor

We were not iterating over all local variables due to a typo.

celinval commented:
r? spastorino

celinval commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -271,6 +323,16 @@ impl<'tcx> Tables<'tcx> {
+
+    fn has_body(&self, instance: Instance<'tcx>) -> bool {
+        let def_id = instance.def_id();
+        !self.tcx.is_foreign_item(def_id)
```
It is valid to create a foreign function instance, at least today. Here's an example in the Miri code: https://github.com/rust-lang/rust/blob/ff0b4b6091a75ad840035a991426c8bc9fbd93bb/src/tools/miri/src/shims/mod.rs#L52. The compiler collector has similar logic, and so does Kani.

compiler-errors replied:
Oh right. Is that not covered by is_mir_available?

celinval replied:
Probably... Let me double check. Btw, in off tomorrow, so I'll get back to it on Friday

celinval replied:
Looking back at this function, I don't know if `is_mir_available()` is the function we need. Here is why, this is only returning the body of local items, where many tools like Kani, MIRI and others also require the body of dependencies, which is accomplished by compiling things with `-Zalways-emit-mir`.

@compiler-errors do you know a better function to use here? I basically want to make sure retrieving the `body()` won't trigger a crash.

celinval commented on compiler/stable_mir/src/mir/mono.rs:
```diff
@@ -54,6 +54,34 @@ impl Instance {
+        with(|cx| cx.resolve_drop_in_place(ty))
+    }
+
+    /// Resolve the drop in place for a given type.
```
Oops...Let me fix that.

celinval commented on compiler/stable_mir/src/mir/mono.rs:
```diff
@@ -54,6 +54,34 @@ impl Instance {
+        })
+    }
+
+    /// Resolve a closure (do we need kind? -- most cases use FnOnce)
```
I'll update the documentation

celinval committed:
Address PR comments

celinval commented:
@compiler-errors let me know if there's anything that you would like me to address. Also, should I reassign this PR for you, or should I wait for @spastorino's review?

celinval committed:
Always run builder to evaluate constants

We were previously skipping it for non-generic functions, but this was
leaving some constants unevaluated.

compiler-errors commented:
r=me after rebasing

celinval committed:
Add more APIs and fix `Instance::body`

Add more APIs to retrieve information about types, and add more instance
resolution options.

Make `Instance::body()` return an Option<Body>, since not every instance
might have an available body. For example, foreign instances, virtual
instances, dependencies.

celinval committed:
Fix has_body() and change resolve_drop_in_place() sig

Fixed the `has_body()` function operator. Before that, this function was
returning false for all shims.

Change resolve_drop_in_place() to also return an instance for empty
shims, since they may still be required for vtable construction.

celinval commented:
@compiler-errors sorry to bother. I added one extra commit to fix 2 APIs. Do you want to take a look at it first?

compiler-errors commented:
I just read them

@bors r+

bors commented:
:pushpin: Commit 8e81fc00879b640ed6acc6a9ca870814189901e5 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 8e81fc00879b640ed6acc6a9ca870814189901e5 -->
<!-- homu: {"type":"Approved","sha":"8e81fc00879b640ed6acc6a9ca870814189901e5","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
